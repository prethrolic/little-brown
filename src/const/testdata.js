export const bookList = [
  {
    "cover": "https://d1w7fb2mkkr3kw.cloudfront.net/assets/images/book/mid/9781/4088/9781408855652.jpg",
    "price": "350",
    "title": "Harry Potter and the Philosopher's Stone (I)",
    "id": "9781408855652"
  },
  {
    "cover": "https://d1w7fb2mkkr3kw.cloudfront.net/assets/images/book/mid/9781/4088/9781408855669.jpg",
    "price": "350",
    "title": "Harry Potter and the Chamber of Secrets (II)",
    "id": "9781408855669"
  },
  {
    "cover": "https://d1w7fb2mkkr3kw.cloudfront.net/assets/images/book/mid/9781/4088/9781408855676.jpg",
    "price": "340",
    "title": "Harry Potter and the Prisoner of Azkaban (III)",
    "id": "9781408855676"
  },
  {
    "cover": "https://d1w7fb2mkkr3kw.cloudfront.net/assets/images/book/mid/9780/2419/9780241975367.jpg",
    "price": "230",
    "title": "The Monk of Mokha",
    "id": "9780241975367"
  },
  {
    "cover": "https://d1w7fb2mkkr3kw.cloudfront.net/assets/images/book/mid/9781/7847/9781784706883.jpg",
    "price": "300",
    "title": "The Language of Kindness",
    "id": "9781784706883"
  },
];

export const cart = {
  "9781408855652": {
    book: {
    "cover": "https://d1w7fb2mkkr3kw.cloudfront.net/assets/images/book/mid/9781/4088/9781408855652.jpg",
    "price": "350",
    "title": "Harry Potter and the Philosopher's Stone (I)",
    "id": "9781408855652"
    },
    count: 2,
  },
  "9781408855669": {
    book: {
      "cover": "https://d1w7fb2mkkr3kw.cloudfront.net/assets/images/book/mid/9781/4088/9781408855669.jpg",
      "price": "350",
      "title": "Harry Potter and the Chamber of Secrets (II)",
      "id": "9781408855669"
    },
    count: 1,
  },
  "9781408855676": {
    book: {
      "cover": "https://d1w7fb2mkkr3kw.cloudfront.net/assets/images/book/mid/9781/4088/9781408855676.jpg",
      "price": "340",
      "title": "Harry Potter and the Prisoner of Azkaban (III)",
      "id": "9781408855676"
    },
    count: 1,
  },
  "9780241975367": {
    book: {
      "cover": "https://d1w7fb2mkkr3kw.cloudfront.net/assets/images/book/mid/9780/2419/9780241975367.jpg",
      "price": "230",
      "title": "The Monk of Mokha",
      "id": "9780241975367"
    },
    count: 2,
  },
};

export const inPromotion = ["9781408855652", "9781408855669", "9781408855676"];

export const user = {
  displayName: 'John Doe',
  pictureUrl: 'https://pbs.twimg.com/profile_images/1269585418107805697/6QKUkoWJ_400x400.jpg',
};

export const paymentInfo = {
  subtotal: '1850.00',
  discount: '114.40',
  total: '1735.60',
  cash: '2000',
};