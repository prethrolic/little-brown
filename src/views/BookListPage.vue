<template>
  <div class='blp-view-container'>
    <Cart />
    <div class='blp-catalog-wrap'>
      <div class='blp-header-container'>
        <h1 class='blp-store-title'>Little Brown Book Shop</h1>
        <div class='blp-profile-container'>
          <div class='blp-profile-picture'>
            <img :src='pictureUrl' />
          </div>
          <div v-html='displayName' class='blp-profile-name' />
          <div class='blp-logout-button' @click='doLogout'>
            Logout
          </div>
        </div>
      </div>
      <div class='blp-tool-container'>
        <div class='blp-search-bar'>
        <div class='icon icon-search' />
        <input
          v-model='keyword'
          type='text'
          placeholder='Search..'
        >
      </div>
      </div>
      <div class='blp-products-container'>
        <div class='blp-products-grid'>
          <BookEntryDisplay
            v-for='book in displayBookList'
            :key='book.id'
            :book='book'
          />
        </div>
      </div>
    </div>
  </div>
</template>
<script>
import BookEntryDisplay from '@/components/BookEntryDisplay.vue';
import Cart from '@/components/Cart.vue';

const PROMOTED_TITLE = 'Harry Potter';

export default {
  name: 'BookListPage',
  components: {
    BookEntryDisplay,
    Cart,
  },
  data() {
    return {
      bookList: [],
      displayBookList: [],
      keyword: '',
      filter: '',
      isTyping: false,
    };
  },
  methods: {
    doLogout() {
      if (this.$liff.isLoggedIn()) {
        this.$liff.logout();
        window.location.reload();
      }
    },
    getBookList() {
      fetch(process.env.VUE_APP_BOOK_LIST_API, {
        method: 'GET',
        contentType: 'application/json',
      })
        .then((res) => res.json())
        .then((data) => {
          const bookList = data.books;
          const newBookList = bookList.map((entry) => {
            const isPromoted = entry.title.toLowerCase().includes(PROMOTED_TITLE.toLowerCase());
            return { ...entry, isPromoted };
          });

          this.bookList = newBookList;
          this.displayBookList = newBookList;
        });
    },
    onBookSearch() {
      this.displayBookList = this.bookList.filter(
        (entry) => entry.title.toLowerCase().includes(this.keyword.toLowerCase()),
      );
    },
  },
  computed: {
    displayName() {
      return this.$store.state.user.displayName;
    },
    pictureUrl() {
      return this.$store.state.user.pictureUrl;
    },
  },
  watch: {
    keyword: {
      handler() {
        if (!this.isTyping) {
          setTimeout(() => {
            this.onBookSearch();
            this.isTyping = false;
          }, 1000);
        }
        this.isTyping = true;
      },
    },
  },
  mounted() {
    this.getBookList();
  },
};
</script>
<style lang="scss" scoped>
@import '@/assets/scss/variables.scss';
@import '@/assets/scss/icons.scss';

.blp {
  &-view-container {
    min-width: 320px;
    width: 100%;
    display: flex;
  }

  &-header-container {
    display: flex;
    align-items: center;
    width: 100%;
    padding: 24px;
  }

  &-store-title {
    margin: 0;
  }

  &-profile-container {
    display: flex;
    align-items: center;
    height: fit-content;
    border: 1px solid $dark-gray;
    border-radius: 12px;
    padding: 8px;
    margin-left: auto;
  }

  &-profile-picture {
    width: 32px;
    height: 32px;
    border-radius: 100%;
    overflow: hidden;
    margin-right: 8px;

    & > img {
      width: 100%;
      height: 100%;
    }
  }

  &-profile-name {
    font-size: 14px;
    font-weight: bold;
    margin-right: 16px;
  }

  &-logout-button {
    display: flex;
    align-items: center;
    justify-content: center;
    padding: 8px 16px;
    background-color: $secondary;
    border-radius: 12px;
    font-size: 12px;
    font-weight: bold;
    color: white;
    cursor: pointer;
    transition: 0.2s ease-in-out;

    &:hover {
      background-color: $secondary-hover;
    }
  }

  &-tool-container {
    padding: 24px;
    align-content: space-between;
  }

  &-search-bar {
    height: fit-content;
    width: fit-content;
    padding: 4px 16px;
    background-color: $light-gray;
    border-radius: 8px;

    & > input {
      display: inline-block;
      width: 25vw;
      height: 32px;
      font-size: 16px;
    }

    & > .icon {
      display: inline-block;
      margin-right: 8px;
      margin-bottom: -2px;
    }
  }

  &-catalog-wrap {
    flex: 1;
    height: 100vh;
    padding: 24px 0;
    display: flex;
    flex-direction: column;
    align-items: flex-start;
    text-align: left;
  }

  &-products-container {
    width: 100%;
    height: 100%;
    overflow-y: scroll;
  }

  &-products-grid {
    width: 100%;
    display: grid;
    grid-auto-rows: 1fr;
    grid-template-columns: 1fr 1fr 1fr;
    grid-gap: 16px;
    padding: 24px;
  }

}
</style>
