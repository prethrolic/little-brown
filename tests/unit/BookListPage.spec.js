import { shallowMount } from '@vue/test-utils';
import BookListPage from '@/views/BookListPage.vue';
import { bookList } from '@/const/testdata.js';


const originalFetch = global.fetch;
const mockJsonPromise = Promise.resolve({ books: bookList });
const mockFetchPromise = Promise.resolve({ json: () => mockJsonPromise });

describe('BookListPage.vue', () => {
  beforeEach(() => {
    global.fetch = jest.fn().mockImplementation(() => mockFetchPromise);
  });

  afterEach(() => {
    global.fetch.mockClear();
    global.fetch = originalFetch;
  });

  it('call fetch correctly', () => {
    const wrapper = shallowMount(BookListPage);
    expect(global.fetch).toHaveBeenCalledTimes(1);
    expect(global.fetch).toHaveBeenCalledWith(process.env.VUE_APP_BOOK_LIST_API, {"contentType": "application/json", "method": "GET"});
  });

  it('render all books', async () => {
    const wrapper = shallowMount(BookListPage);
    await wrapper.setData({
      displayBookList: bookList,
    })
    
    expect(wrapper.element).toMatchSnapshot();
  });
});
