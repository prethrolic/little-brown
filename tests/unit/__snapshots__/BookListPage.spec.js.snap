// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BookListPage.vue render all books 1`] = `
<div
  class="blp-view-container"
>
  <div>
    <div
      class="blp-cart-button"
    >
      <div
        class="icon-cart-32"
      />
       
      <!---->
    </div>
  </div>
   
  <div
    class="blp-catalog-wrap"
  >
    <div
      class="blp-header-container"
    >
      <h1
        class="blp-store-title"
      >
        Little Brown Book Shop
      </h1>
       
      <div
        class="blp-profile-container"
      >
        <div
          class="blp-profile-picture"
        >
          <img
            src="https://pbs.twimg.com/profile_images/1269585418107805697/6QKUkoWJ_400x400.jpg"
          />
        </div>
         
        <div
          class="blp-profile-name"
        >
          John Doe
        </div>
         
        <div
          class="blp-logout-button"
        >
          
          Logout
        
        </div>
      </div>
    </div>
     
    <div
      class="blp-tool-container"
    >
      <div
        class="blp-search-bar"
      >
        <div
          class="icon icon-search"
        />
         
        <input
          placeholder="Search.."
          type="text"
        />
      </div>
    </div>
     
    <div
      class="blp-products-container"
    >
      <div
        class="blp-products-grid"
      >
        <bookentrydisplay-stub
          book="[object Object]"
        />
        <bookentrydisplay-stub
          book="[object Object]"
        />
        <bookentrydisplay-stub
          book="[object Object]"
        />
        <bookentrydisplay-stub
          book="[object Object]"
        />
        <bookentrydisplay-stub
          book="[object Object]"
        />
      </div>
    </div>
  </div>
</div>
`;
